File: src/main/java/net/oxyopia/vice/utils/DevUtils.java
package net.oxyopia.vice.utils;

import gg.essential.universal.UChat;
import net.minecraft.client.MinecraftClient;
import net.minecraft.text.ClickEvent;
import net.minecraft.text.HoverEvent;
import net.minecraft.text.Style;
import net.minecraft.text.Text;
import net.oxyopia.vice.Vice;
import net.oxyopia.vice.config.DevConfig;
import java.lang.reflect.Field;

import static net.oxyopia.vice.Vice.*;

public class DevUtils {
	public static void sendDebugChat(String msg) {
		if (Vice.config.DEVMODE) {
			UChat.chat(Vice.DEV_PREFIX + msg.replaceAll("&&", "§"));
		}
	}

	public static void sendDebugChat(String msg, String fieldName) {
		if (!Vice.config.DEVMODE) return;

		try {
			Field field = DevConfig.class.getDeclaredField(fieldName);
			field.setAccessible(true);
			boolean value = (boolean) field.get(devConfig);

			if (value) {
				sendDebugChat(msg);
			}
		} catch (NoSuchFieldException | IllegalAccessException e) {
			sendErrorMessage(e, "An error occurred sending a debug chat (really meta i know)!");
		}
	}

	public static void sendErrorMessage(Throwable throwable, String msg) {
		StackTraceElement[] stackTrace = throwable.getStackTrace();
		StringBuilder stackTraceString = new StringBuilder(throwable.getMessage());

		for (StackTraceElement element : stackTrace) {
			stackTraceString.append("\n\t").append(element.toString());
		}

		String joinedStackTrace = stackTraceString.toString();

		Text errorChat = Text.literal(ERROR_PREFIX + msg + "§7 Click to copy the error to clipboard!")
			.setStyle(Style.EMPTY
				.withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, joinedStackTrace))
				.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, Text.literal("§7Click to copy the error to your clipboard!\n§c"+msg))));

		Vice.Companion.getLogger().error(throwable.getMessage(), throwable);
		if (MinecraftClient.getInstance().inGameHud.getChatHud() != null) {
			MinecraftClient.getInstance().inGameHud.getChatHud().addMessage(errorChat);
		}
	}

	public static void sendWarningMessage(String msg) {
		msg = msg.replaceAll("&&", "§");
		MinecraftClient.getInstance().inGameHud.getChatHud().addMessage(Text.of(WARNING_PREFIX + msg));
	}

	public static void sendWarningMessage(String msg, String copyableData) {
		msg = msg.replaceAll("&&", "§");

		Text warningChat = Text.literal(WARNING_PREFIX + msg)
			.setStyle(Style.EMPTY
				.withClickEvent(new ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, copyableData))
				.withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, Text.literal("§eClick to copy some useful data to your clipboard!")))
			);

		MinecraftClient.getInstance().inGameHud.getChatHud().addMessage(warningChat);
	}
}

File: src/main/java/net/oxyopia/vice/utils/DevUtils.kt
package net.oxyopia.vice.utils

import net.minecraft.client.MinecraftClient
import net.minecraft.text.ClickEvent
import net.minecraft.text.HoverEvent
import net.minecraft.text.Style
import net.minecraft.text.Text
import net.oxyopia.vice.Vice
import net.oxyopia.vice.config.DevConfig
import net.oxyopia.vice.utils.Utils.convertFormatting

object DevUtils {
	fun sendDebugChat(msg: String) {
		if (!Vice.config.DEVMODE) return
		MinecraftClient.getInstance().inGameHud.chatHud.addMessage(Text.of(Vice.DEV_PREFIX + msg.convertFormatting()))
	}

	@JvmStatic
	fun sendDebugChat(msg: String, fieldName: String) {
		if (!Vice.config.DEVMODE) return

		try {
			val field = DevConfig::class.java.getDeclaredField(fieldName) ?: return
			field.isAccessible = true
			val value = field.getBoolean(Vice.devConfig)

			if (value) {
				sendDebugChat(msg)
			}
		} catch (e: NoSuchFieldException) {
			sendErrorMessage(e, "An error occurred sending a debug chat (really meta i know)!")
		} catch (e: IllegalAccessException) {
			sendErrorMessage(e, "An error occurred sending a debug chat (really meta i know)!")
		}
	}

	fun sendErrorMessage(throwable: Throwable, msg: String) {
		val stackTrace = throwable.stackTrace
		val stackTraceString = StringBuilder(throwable.message)

		for (element in stackTrace) {
			stackTraceString.append("\n\t").append(element.toString())
		}

		val joinedStackTrace = stackTraceString.toString()

		val errorChat: Text = Text.literal(Vice.ERROR_PREFIX + msg + "§7 Click to copy the error to clipboard!")
			.setStyle(
				Style.EMPTY
					.withClickEvent(ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, joinedStackTrace))
					.withHoverEvent(
						HoverEvent(
							HoverEvent.Action.SHOW_TEXT,
							Text.literal("§7Click to copy the error to your clipboard!\n§c$msg")
						)
					)
			)

		Vice.logger.error(throwable.message, throwable)
		MinecraftClient.getInstance().inGameHud.chatHud.addMessage(errorChat)
	}

	fun sendWarningMessage(msg: String) {
		MinecraftClient.getInstance().inGameHud.chatHud.addMessage(Text.of(Vice.WARNING_PREFIX + msg.convertFormatting()))
	}

	fun sendWarningMessage(msg: String, copyableData: String?) {
		val warningChat: Text = Text.literal(Vice.WARNING_PREFIX + msg.convertFormatting())
			.setStyle(
				Style.EMPTY
					.withClickEvent(ClickEvent(ClickEvent.Action.COPY_TO_CLIPBOARD, copyableData))
					.withHoverEvent(
						HoverEvent(
							HoverEvent.Action.SHOW_TEXT,
							Text.literal("§eClick to copy some useful data to your clipboard!")
						)
					)
			)

		MinecraftClient.getInstance().inGameHud.chatHud.addMessage(warningChat)
	}
}
